tools/c7n_azure/c7n_azure/resources/access_control.py

```diff
@@ -96,3 +141,112 @@ class UserRole(RelatedResourceFilter):
 
     RelatedResource = "c7n_azure.resources.access_control.RoleDefinition"
     RelatedIdsExpression = "properties.roleDefinitionId"
+
+
+@RoleAssignment.filter_registry.register('resource-access')
+class ResourceAccessFilter(RelatedResourceFilter):
+    """Filters role assignments that have access to a certain
+    type of azure resource.
+
+    :Example:
+
+        .. code-block:: yaml
+
+            policies:
+               - name: assignments-by-azure-resource
+                 resource: azure.roleassignment
+                 filters:
+                    - type: resource-access
+                      relatedResource: azure.vm
+    """
+
+    schema = type_schema(
+        'resource-access',
+        relatedResource={'type': 'string'},
+        rinherit=RelatedResourceFilter.schema,
+        required=['relatedResource']
+    )
+
+    def __init__(self, data, manager=None):
+        super(ResourceAccessFilter, self).__init__(data, manager)
+        resource_type = self.data['relatedResource']
+        self.factory = Azure.resources.get(
+            resource_type.rsplit('.', 1)[-1])
+
+    def get_related_ids(self, resources, related):
+        related_ids = []
+        for r in related:
+            for resource in resources:
+                if resource['properties']['scope'] in r:
+                    related_ids.append(r)
+                    continue
+
+        return related_ids
+
+    def get_related(self, resources):
+        ctx = ExecutionContext(local_session(Session), self.data, Config.empty())
+        manager = self.factory(ctx, self.data)
+        related = manager.source.get_resources(None)
+        return {r['id']: r for r in related}
+
+    def process_resource(self, resource, related):
+        if not self.data.get('key'):
+            return True
+
+        related_ids = self.get_related_ids([resource], related)
+        model = self.manager.get_model()
+        op = self.data.get('operator', 'or')
+        found = []
+        if self.data.get('match-resource') is True:
+            self.data['value'] = self.get_resource_value(
+                self.data['key'], resource)
+        for rid in related_ids:
```

Comment:
Really everything you are doing in this filter is stemming from this (normally unreachable) code that probably shouldn't even be here :)  Process resources shouldn't need to call or look at `related_ids` because `related` is already filtered resource set from `get_related()`.  

I guess you can remove all references to ID's from this method and just iterate over `related` instead, then you don't need quite so much code (you could delete your `get_related_ids` implementation and just update `get_related` to do a contains instead of an equality check on the last line)

Tempting to just remove this unnecessary test from the base class :)

Reply:
I don't think this is typically unreachable though, unless I don't understand which line you are commenting on. I actually hit this code every time I've done a RelatedResourceFilter that actually had a related id on the resource. 

What is happening in this code is that it gets all the specific related resources for the ONE resource being processed, then applies the value filter on the related resource.  If the related resource fulfills its ValueFilter then the actual resource is not filtered out. 

But the same thing could be done by just applying the value filter on all the related resources when it gets them and only returns those related resources, and then filters down the original resources based on the related ids.

Reply:
A more simple one to understand this code is the role assignment and role definition related filter. This code gets the specific RoleDefinition for the RoleAssignment being processed, applies the Value filter (e.g. Owner) to the related resource (Role Definition), and then filters everyone else where the related role definition name isn't Owner: 

```
policies:
   - name: assignment-by-role
     resource: azure.roleassignment
     filters:
        - type: role
          key: properties.roleName
          op: eq
          value: Owner
```